File: src/main/java/com/back/together02be/ranking/controller/RankingController.java
package com.back.together02be.ranking.controller;

import com.back.together02be.ranking.dto.response.RankingRes;
import com.back.together02be.ranking.service.RankingSeasonService;
import com.back.together02be.ranking.service.RankingService;
import com.back.together02be.ranking.service.RankingSnapshotService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.util.List;

@RestController
@RequestMapping("/api/rankings")
@RequiredArgsConstructor
public class RankingController {

    private final RankingService rankingService;
    private final RankingSnapshotService rankingSnapshotService;
    private final RankingSeasonService rankingSeasonService;

    // 메인 랭킹 조회 API이다.
    @GetMapping
    public List<RankingRes> getRankings() {
        return rankingService.getDailyRankings();
    }

    // DAILY 랭킹을 생성한다.
    @PostMapping("/snapshots/daily")
    public String createDailySnapshot(@RequestParam(required = false) LocalDate snapshotDate) {
        LocalDate targetDate = (snapshotDate != null) ? snapshotDate : LocalDate.now();
        rankingSnapshotService.createDailySnapshot(targetDate);
        return "DAILY 랭킹 생성 완료: " + targetDate;
    }

    // MONTHLY 랭킹을 생성한다.
    @PostMapping("/snapshots/monthly")
    public String createMonthlySnapshot(@RequestParam LocalDate snapshotDate) {
        rankingSnapshotService.createMonthlySnapshot(snapshotDate);
        return "MONTHLY 랭킹 생성 완료: " + snapshotDate;
    }

    // 특정 날짜의 MONTHLY 랭킹을 조회한다.
    @GetMapping("/monthly")
    public List<RankingRes> getMonthlyRankings(@RequestParam LocalDate snapshotDate) {
        return rankingService.getMonthlyRankings(snapshotDate);
    }

    // 시즌 시작 시 기준 자산을 생성한다.
    @PostMapping("/season/start")
    public String startSeason(@RequestParam(required = false) LocalDate startDate) {
        LocalDate targetDate = (startDate != null) ? startDate : LocalDate.now();
        rankingSeasonService.startSeason(targetDate);
        return "시즌 시작 완료: " + targetDate;
    }

    // 시즌 종료 후 다음 시즌 기준 자산을 다시 생성한다.
    @PostMapping("/season/reset")
    public String resetSeason(
            @RequestParam LocalDate endDate,
            @RequestParam LocalDate nextStartDate
    ) {
        rankingSeasonService.resetSeason(endDate, nextStartDate);
        return "시즌 리셋 완료: " + endDate + " -> " + nextStartDate;
    }
}

File: src/main/kotlin/com/back/together02be/ranking/controller/RankingController.kt
package com.back.together02be.ranking.controller

import com.back.together02be.ranking.dto.response.RankingRes
import com.back.together02be.ranking.service.RankingSeasonService
import com.back.together02be.ranking.service.RankingService
import com.back.together02be.ranking.service.RankingSnapshotService
import org.springframework.web.bind.annotation.*
import java.time.LocalDate

@RestController
@RequestMapping("/api/rankings")
class RankingController(
    private val rankingService: RankingService,
    private val rankingSnapshotService: RankingSnapshotService,
    private val rankingSeasonService: RankingSeasonService
) {

    @GetMapping
    fun getRankings(): List<RankingRes> {
        return rankingService.getDailyRankings()
    }

    @PostMapping("/snapshots/daily")
    fun createDailySnapshot(@RequestParam(required = false) snapshotDate: LocalDate?): String {
        val targetDate = snapshotDate ?: LocalDate.now()
        rankingSnapshotService.createDailySnapshot(targetDate)

        // 코틀린의 문자열 템플릿($)을 활용해 + 연산자를 제거
        return "DAILY 랭킹 생성 완료: $targetDate"
    }

    @PostMapping("/snapshots/monthly")
    fun createMonthlySnapshot(@RequestParam snapshotDate: LocalDate): String {
        rankingSnapshotService.createMonthlySnapshot(snapshotDate)
        return "MONTHLY 랭킹 생성 완료: $snapshotDate"
    }

    @GetMapping("/monthly")
    fun getMonthlyRankings(@RequestParam snapshotDate: LocalDate): List<RankingRes> {
        return rankingService.getMonthlyRankings(snapshotDate)
    }

    @PostMapping("/season/start")
    fun startSeason(@RequestParam(required = false) startDate: LocalDate?): String {
        val targetDate = startDate ?: LocalDate.now()
        rankingSeasonService.startSeason(targetDate)
        return "시즌 시작 완료: $targetDate"
    }

    @PostMapping("/season/reset")
    fun resetSeason(
        @RequestParam endDate: LocalDate,
        @RequestParam nextStartDate: LocalDate
    ): String {
        rankingSeasonService.resetSeason(endDate, nextStartDate)
        return "시즌 리셋 완료: $endDate -> $nextStartDate"
    }
}